spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RecurrentDecoder.scala

```diff
@@ -0,0 +1,217 @@
+    outputSize(1) = hiddenSize
+    require(hiddenSize == input.size()(1), "hiddenSize is " +
+      "not the same with input size!! Please update cell settings or use Recurrent instead!")
+    val featureSizes = outputSize.drop(1)
+    output.resize(Array(batchSize, times) ++ featureSizes)
+    // Clone N modules along the sequence dimension.
+    extend(featureSizes)
+    if (preTopology != null) newInput.resize(output.size())
+    else outputCell.resize(output.size())
+
+    /**
+     * currentInput forms a T() type. It contains two elements, hidden and input.
+     * Each time it will feed the cell with T(hidden, input) (or T(input, hidden) depends on
+     * your hidDim and inputDim), and the cell will give a table output containing two
+     * identical elements T(output, output). One of the elements from the cell output is
+     * the updated hidden. Thus the currentInput will update its hidden element with this output.
+     */
+    var i = 1
+    // init state
+    currentInput(hidDim) = if (initState != null) initState
+     else hidden
+
+    while (i <= times) {
+      // input at t(0) is user input
+      val inputTmp = if (i == 1) {
+        input
+      } else {
+        // input at t(i) is output at t(i-1)
+        cells(i - 2).output.toTable[Tensor[T]](inputDim)
+      }
+
+      currentInput(inputDim) = if (preTopology != null) {
+        newInput.narrow(2, i, 1).copy(inputTmp)
+        val sizes = 1 +: inputTmp.size()
+        inputTmp.resize(sizes)
+        val _input = preTopology.updateOutput(inputTmp).toTensor[T]
+        inputTmp.resize(sizes.takeRight(sizes.length - 1))
+        _input.select(1, 1)
+      } else {
+        outputCell.narrow(2, i, 1).copy(inputTmp)
+        inputTmp
+      }
+      cells(i - 1).updateOutput(currentInput)
+      currentInput(hidDim) = cells(i - 1).output.toTable(hidDim)
+      i += 1
+    }
+
+    if (preTopology != null) {
+      // For backward preTopology use
+      outputCell = preTopology.updateOutput(newInput).toTensor[T]
+    }
+
+    Recurrent.copy(cells.map(x => x.output.toTable[Tensor[T]](inputDim)), output)
+    output
+  }
+
+  override def accGradParameters(input: Tensor[T], gradOutput: Tensor[T]): Unit = {
+    throw new Exception("Should not enter RecurrentDecoder accGradParameters" +
+      "as it has override backward")
+  }
+
+  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    throw new Exception("Should not enter RecurrentDecoder updateGradInput" +
+      "as it has override backward")
+    gradInput
+  }
+
+  override def backward(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    val st = System.nanoTime
+    currentGradOutput(hidDim) = gradHidden
+    var i = times
+    while (i >= 1) {
+      currentGradOutput(inputDim) = if (i == times) gradOutput.select(timeDim, i)
+      else {
+        gradOutput.select(timeDim, i).clone()
+          .add(cells(i).gradInput.toTable[Tensor[T]](inputDim).clone())
```

Comment:
Use Recurrent.selectCopy instead of gradOutput.select, and then you can directly add cells(i).gradInput.toTable[Tensor[T]](inputDim); no need to clone the latter